### ISSUE TITLE
helm:Make image tags/name/repository configurable separately

### ISSUE THREAD

**Issue by cucxabong** [hashicorp/consul-k8s#644]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request. Searching for pre-existing feature requests helps us consolidate datapoints for identical requirements into a single place, thank you!
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.

---

 We have an in-house tool that will fetch latest tags for an image in a (local docker registry) & then fill values (using `--set`) file on deployment, so this tool can not work when current chart requires input image as a single string (with registry/name/tag in a single line)

#### Contributions

I'm happy to work on this

**Comment by lkysow** [comment]
I don't think we can support this without breaking backwards compatibility. Say we do as @thisisnotashwin suggests and have:

```yaml
# values.yaml
global:
  imageRepo: docker.io
  imageK8S: hashicorp/consul-k8s:0.36.0
```

And then in our templates we use:

```yaml
image: {{ .Values.global.imageRepo }}/{{ .Values.imageK8S}}
```

This would break anyone that was previously setting the full image, e.g.:

```yaml
global:
  imageK8S: my.internal.repo/hashicorp/consul-k8s:0.36.0
```

Since it would now render as:

```yaml
image: docker.io/my.internal.repo/hashicorp/consul-k8s:0.36.0
```

Given that this is a nice-to-have and that there should be a way to work around it in tooling and that I don't see a way to implement this without breaking backwards compatibility I'm not sure it makes sense to implement this.

**Comment by TJM** [comment]
Without breaking backwards compatibility... If the parameter to set the "full image name" (global.imageK8S) is set, use it, otherwise construct an image name using the component parts. Just don't use the same parameter for the full image name as  the component parts. The "image" that is passed to the kubernetes resource can all be magic'd out in a helpers template.

**Comment by ebachle** [comment]
hey @david-yu I wanted to bump this one.  Looking at the chart I can see how this would be a challenging refactor, but I do want to highlight how worthwhile my organization would find it.  

In our case, we're angling to get full off of Dockerhub and onto the Public ECR Gallery for images so we don't leave the AWS network to pull images.

In the case of the vault chart, it was fairly simple to make this migration.  We replaced the repository on each component and we were good to go.  

https://github.com/hashicorp/vault-helm/blob/main/values.yaml#L68-L71

```
  image:
    repository: "hashicorp/vault-k8s"
    tag: "1.2.1"
```

With consul however, the part that somewhat holds me back from taking the leap is that I would then end up having to take up tracing both the versions of the chart and the images.  While that is doable and if I want off of Dockerhub bad enough, I'll probably do it.  But it puts me in a position of not just worrying about my chart version and what the associated default version of Consul is for that chart, but matching those up and keeping them up-to-date.

From my end, it would definitely be preferable if I was able to update the repository without having to take up the additional task of keeping an eye on the version of the image as a side-effect.

My thought on a possible contract that might work would be to do something like Vault does, but test for if `image` is set to a string or a map in a template helper and if it's a string, take that value, otherwise compose it from the map.  That might make it less breaking, though not less work from an integration testing side.

Thanks for the consideration!

**Comment by dmpe** [comment]
related https://github.com/hashicorp/consul-k8s/issues/3425

**Comment by ebachle** [comment]
I also found myself itching for this one again as the 1.16.7 image didn't make it to dockerhub, so I was considering changing the registry.  

Though the core fix probably needs to be to ensure the images get published.  

Ref: https://hub.docker.com/r/hashicorp/consul/tags?page=&page_size=&ordering=&name=1.16
